web/src/narrow.js

```diff
@@ -94,6 +94,56 @@ export function save_narrow(terms, trigger) {
     changehash(new_hash, trigger);
 }
 
+function change_view_for_narrow_containing_with_operator(message_data, opts) {
```

Comment:
I'm not sure I see anywhere where this logic would avoid doing a bunch of work in the very common case that the target message has never moved. Possibly the `message_fetch` API response suggested elsewhere would be helpful to you here for doing this check well.

Reply:
I added the logic so that it would avoid the work in case it is in the same narrow.